stackdriver/config/config.go

```diff
@@ -0,0 +1,128 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"google.golang.org/api/option"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// ConfigMapNameEnv is the name of the environment variable that contains the name of
+	// the ConfigMap for stackdriver configuration.
+	ConfigMapNameEnv = "CONFIG_STACKDRIVER_NAME"
+	// secretNamespace is the namespace in which secrets for authenticating with Stackdriver
+	// should be stored.
+	secretNamespace = "knative-serving"
+	// secretDataFieldKey is key of the Kubernetes Secret Data field that contains the GCP
+	// service account key to use.
+	secretDataFieldKey = "key.json"
+
+	// ProjectIDKey is the name of the ConfigMap field for ProjectID.
+	ProjectIDKey = "project-id"
+	// GcpLocationKey is name of the ConfigMap field for GcpLocation.
+	GcpLocationKey = "gcp-location"
+	// ClusterNameKey is the name of the ConfigMap field for ClusterName.
+	ClusterNameKey = "cluster-name"
+	// GcpSecretNameKey is the name of the ConfigMap field for GcpSecretName.
+	GcpSecretNameKey = "gcp-secret-name"
+)
+
+// Config encapsulates the metadata required to configure a Stackdriver client.
+type Config struct {
+	// ProjectID is the stackdriver project ID to which data is uploaded.
+	// This is not necessarily the GCP project ID where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ProjectID string
+	// GcpLocation is the GCP region or zone to which data is uploaded.
+	// This is not necessarily the GCP project ID where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	GcpLocation string
+	// ClusterName is the cluster name with which the data will be associated in Stackdriver.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ClusterName string
+	// GcpSecretName is the optional GCP service account key which will be used to
+	// authenticate with Stackdriver. If not provided, Google Application Default Credentials
+	// will be used (https://cloud.google.com/docs/authentication/production).
+	GcpSecretName string
+}
+
+// NewStackdriverConfigFromConfigMap returns a Config for the given configmap
+func NewStackdriverConfigFromConfigMap(config *corev1.ConfigMap) (*Config, error) {
+	return NewStackdriverConfigFromMap(config.Data)
+}
+
+// NewStackdriverConfigFromMap returns a Config for the given map
+func NewStackdriverConfigFromMap(config map[string]string) (*Config, error) {
```

Comment:
I would expect unit tests for funcs in this file.

Reply:
On the way!